docs/contributing.rst

```diff
@@ -0,0 +1,201 @@
+    [GCC 6.3.0 20170321] on linux
+
+3. An explanation of why the current behaviour is wrong/not desired, and what you
+   expect instead.
+
+Enhancement proposals
+---------------------
+
+If you have an idea about a new feature or some other improvement to Zarr, please raise a
+`GitHub issue <https://github.com/zarr-developers/zarr/issues/new>`_ first to discuss.
+
+We very much welcome ideas and suggestions for how to improve Zarr, but please bear in
+mind that we are likely to be conservative in accepting proposals for new features. The
+reasons for this are that we would like to keep the Zarr code base lean and focused on
+a core set of functionalities, and available time for development, review and maintenance
+of new features is limited. But if you have a great idea, please don't let that stop
+you posting it on GitHub, just please don't be offended if we respond cautiously.
+
+Contributing code and/or documentation
+--------------------------------------
+
+Forking the repository
+~~~~~~~~~~~~~~~~~~~~~~
+
+The Zarr source code is hosted on GitHub at the following location:
+
+* `https://github.com/zarr-developers/zarr <https://github.com/zarr-developers/zarr>`_
+
+You will need your own fork to work on the code. Go to the link above and hit
+the "Fork" button. Then clone your fork to your local machine::
+
+    $ git clone git@github.com:your-user-name/zarr.git
+    $ cd zarr
+    $ git remote add upstream git@github.com:zarr-developers/zarr.git
+
+Creating a development environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To work with the Zarr source code, it is recommended to set up a Python virtual
+environment and install all Zarr dependencies using the same versions as are used by
+the core developers and continuous integration services. Assuming you have a Python
+3 interpreter already installed, and have also installed the virtualenv package, and
+you have cloned the Zarr source code and your current working directory is the root of
+the repository, you can do something like the following::
+
+    $ mkdir -p ~/pyenv/zarr-dev
+    $ virtualenv --no-site-packages --python=/usr/bin/python3.6 ~/pyenv/zarr-dev
+    $ source ~/pyenv/zarr-dev/bin/activate
+    $ pip install -r requirements_dev.txt
+    $ pip install -r requirements_dev_optional.txt
+    $ pip install -e .
+
+To verify that your development environment is working, you can run the unit tests::
+
+    $ pytest -v zarr
+
+Creating a branch
+~~~~~~~~~~~~~~~~~
+
+It's best to create a new, separate branch for each piece of work you want to do. E.g.::
+
+    git fetch upstream
+    git checkout -b shiny-new-feature upsteam/master
+
+This changes your working directory to the shiny-new-feature branch. Keep any changes in
+this branch specific to one bug or feature so it is clear what the branch brings to
+Zarr.
+
+To update this branch with latest code from Zarr, you can retrieve the changes from
+the master branch::
+
+    git fetch upstream
+    git rebase upstream/master
```

Comment:
While I love rebasing personally, working with developers from various skill levels has caused me to suggest merging to more general audiences. Rebasing can mean trying to resolve many conflicts in many steps instead of all conflicts in one step. The latter seems to avoid many new issues cropping up and allows any new issues to be resolved in subsequent commits pretty easily. Not to mention merging tends to be friendlier if other people help out on your branch.

These are all personal opinions and observations over time. Feel free to ignore them if you have a strong preference.

Reply:
It's a fair point, I've seen people struggle with rebase, I remember the first time I tried it I went round in circles for about an hour before someone told me you had to do ``git push -f`` at the end.